rfcs/0097-no-read-store-dir.md

```diff
@@ -0,0 +1,119 @@
+etc., thus this change is only important for sandboxes where you can remove such information leaks.
+It isn't a big improvement by itself, but it is a small incremental hardening with few drawbacks that allows for better security in combination with sandboxes.
+
+NB: While this change is simple, it is not possible for end-users to do without changing Nix, since at the moment any
+`nix` command will reset the permissions back to 1775.
+
+# Detailed design
+[design]: #detailed-design
+
+Set the 1775
+in [nixpkgs/nixos/modules/system/boot/stage-2-init.sh](https://github.com/NixOS/nixpkgs/blob/8284fc30c84ea47e63209d1a892aca1dfcd6bdf3/nixos/modules/system/boot/stage-2-init.sh#L62),
+in [nix/scripts/install-multi-user.sh](https://github.com/NixOS/nix/blob/cf1d4299a8fa8906f62271dcd878018cef84cc30/scripts/install-multi-user.sh#L577),
+in [nix/src/libstore/globals.hh](https://github.com/NixOS/nix/blob/ba8b39c13003c8ddafb6bec308997e09b9851c46/src/libstore/globals.hh#L278),
+in [nix/src/libstore/build/local-derivation-goal.cc](https://github.com/NixOS/nix/blob/6182ae689826554d915b4ed72e07f7978dc1d13c/src/libstore/build/local-derivation-goal.cc#L641), and
+in [nix/src/libstore/local-store.cc](https://github.com/NixOS/nix/blob/0a535dd5ac93576f7152d786464e330ae3d46b50/src/libstore/local-store.cc#L181)
+to 1771.
+
+/nix/store/trash and /nix/store/.links will also have to have their read bit removed, resulting in 0751.
+
+Currently, /proc/cmdline provides the path to the current system configuration, which is counter-productive in this case.
+This would be fixed by simply setting the permissions to 0440, since the permissions are universal to all pid namespaces.
+Fixing the permissions for other files in /proc would likely also be a good idea, regardless of this RFC and regardless
+of sandboxing, e.g. it doesn't make sense to expose /proc/config.gz to the entire world even if it doesn't expose any secrets.
+
+Nothing else has to be done likely, since setting the store permissions to `1771` manually doesn't break
+anything other than what is mentioned in this document (though it is undone if you run a `nix` command).
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+Losing the read (r) bit means that you can't list the files inside the store.
+The execute (x) bit allows us to `cd` to it and also access paths inside the store.
+
+E.g. `ls "$(readlink /nix/var/nix/profiles/system)"` will still work, since this is a directory
+inside the store, and not the store itself, but an unprivileged user can't `ls /nix/store` to find the system configuration.
+
+Note: A sandboxed program could still have some idea of how the host machine is used by checking each store path that Hydra has ever built,
+then it could estimate whether the host machine has e.g. Tor on it.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+It might be a slight annoyance since shell completion won't work in the /nix/store anymore, e.g.
```

Comment:
I am constantly relying on shell completion of hashes to investigate what is going on in NixOS, so I view this as a major step back in usability. Maybe there's some other way I could be working, like completing paths base on shell scrollback instead of path completion?

Reply:
I personally just copy the path in foot (my terminal) by pressing Ctrl+Shift+R.

Reply:
I think I even manage to have workflows that go both ways — sometimes it is nice to complete from a few letters, nicer than doing scrollback search then needing to do selection (or taking the mouse); sometimes it is less nice when zsh rechecks whether intermediate components in the path need expansion on Tab completion

Reply:
You ought to use foot if possible, it's a great terminal, you basically select what to copy by searching for a prefix of it.

Reply:
That sounds basically what screen does anyway, maybe with more automagic for detecting the end, which will probably sometimes go wrong, even if not too often…

Nah, not even on the same scale as having a chance to affect my willingness of giving up my window manager configuration. (Then again, I don't even use urxvt-specific stuff much and sometimes end up working in fbterm…)